### PR TITLE
feat(infra): use SAML identity provider for users auth

### PR DESCRIPTION
**Pull request by imincik**
This change is introducing Data Lake users authorization using SAML identity provider . It is replacing current authorization based on AWS account ID.

SAML auth is successfully tested manually.

## Issues

Closes: https://github.com/linz/geospatial-data-lake-team/issues/39

### CODE REVIEW

File: USAGE.md [hunk 1]
```diff
@@ -146,7 +146,7 @@ aws lambda invoke \
 ## Dataset Version Endpoint Usage Examples
 
 - Set Dataset Space Endpoint Lambda function name
-  
+
```
Comment (l0b0) [inline USAGE.md:149]: I believe these newlines will break some Markdown renderers. If the original Markdown isn't rendering properly somewhere, please let me know. Also, all the whitespace changes in this commit probably belongs in a different PR.

File: infrastructure/users_stack.py [hunk 2]
```diff
@@ -11,21 +11,25 @@ class UsersStack(Stack):
     def __init__(self, scope: Construct, stack_id: str, **kwargs: Any) -> None:
         super().__init__(scope, stack_id, **kwargs)
 
-        account_ids = (
-            aws_iam.AccountPrincipal(account_id=account_id)
-            for account_id in os.environ.get(
-                "DATALAKE_USERS_AWS_ACCOUNTS_IDS", aws_iam.AccountRootPrincipal().account_id
-            ).split(",")
-        )
+        saml_provider_arn = os.environ.get("DATALAKE_SAML_IDENTITY_PROVIDER_ARN", None)
```
Comment (l0b0) [inline infrastructure/users_stack.py:14]: The `, None` part isn't necessary.

File: USAGE.md [hunk 3]
```diff
@@ -5,7 +5,7 @@ The purpose of GDL is to store geospatial datasets. This document should provide
 The keywords "must", "must not", "required", "shall", "shall not", "should", "should not", "recommended",  "may", and "optional" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
 
 ## Prerequisites
-Currently, Data Lake allows read/write access for all users from all AWS accounts (users home accounts) specified in `DATALAKE_USERS_AWS_ACCOUNTS_IDS` during deployment time (see [README](README.md#aws-infrastructure-deployment-cdk-stack)). Please contact the Data Lake product team to add your AWS account to this list.
+Currently, Data Lake allows read/write access for all users authenticated via `DATALAKE_SAML_IDENTITY_PROVIDER_ARN` during deployment time (see [README](README.md#aws-infrastructure-deployment-cdk-stack)).
```
Comment (billgeo) [inline USAGE.md:8]: Do we need to add some info on how to get added to the AD group, like request access from the LINZ service desk? The name of the group? Not sure that we should put this in a public repo though, as a) it's not relevant to anyone outside of LINZ and b) it could be some minor security risk? Could put it in Confluence?

File: infrastructure/users_stack.py [hunk 4]
```diff
@@ -11,21 +11,25 @@ class UsersStack(Stack):
...
+            principal = aws_iam.AccountPrincipal(  # type: ignore[assignment]
+                account_id=aws_iam.AccountRootPrincipal().account_id
+            )
 
         self.users_role = aws_iam.Role(
             self,
             "users-role",
             role_name=ResourceName.USERS_ROLE_NAME.value,
-            assumed_by=principals,  # type: ignore[arg-type]
+            assumed_by=principal,  # type: ignore[arg-type]
```
Comment (l0b0) [inline infrastructure/users_stack.py:32]: Is the `# type: ignore[arg-type]` annotation superfluous now?

**Review by billgeo — CHANGES_REQUESTED** [review]

**Review by l0b0 — CHANGES_REQUESTED** [review]

File: USAGE.md (hunk 3 above)
Comment (billgeo) [reply USAGE.md:8]: > Data Lake allows read/write access for all users authenticated via `DATALAKE_SAML_IDENTITY_PROVIDER_ARN` during deployment time

This seems more like implementation detail rather than how to use it.

Comment (imincik) [reply USAGE.md:8]: > Do we need to add some info on how to get added to the AD group, like request access from the LINZ service desk? The name of the group? Not sure that we should put this in a public repo though, as a) it's not relevant to anyone outside of LINZ and b) it could be some minor security risk? Could put it in Confluence?

Yes, this information is not relevant for public. It should go to Confluence.

Comment (imincik) [reply USAGE.md:8]: How to use `DATALAKE_SAML_IDENTITY_PROVIDER_ARN` as deployment configuraion is explained in `README`.

File: USAGE.md (hunk 1 above)
Comment (imincik) [reply USAGE.md:149]: They must have been done by some automatic formatting in my editor. I'll remove them.

Comment (imincik) [reply USAGE.md:149]: Actually, it is not new lines but removed two spaces at the beginning of new lines.

File: infrastructure/users_stack.py (hunk 2 above)
Comment (imincik) [reply infrastructure/users_stack.py:14]: Removed.

File: infrastructure/users_stack.py (hunk 4 above)
Comment (imincik) [reply infrastructure/users_stack.py:32]: No, when removed I get following error:

```
infrastructure/users_stack.py:32: error: Argument "assumed_by" to "Role" has incompatible type "FederatedPrincipal"; expected "IPrincipal"  [arg-type]
infrastructure/users_stack.py:32: note: Following member(s) of "FederatedPrincipal" have conflicts:
infrastructure/users_stack.py:32: note:     Expected:
infrastructure/users_stack.py:32: note:         def __jsii_proxy_class__() -> Type[_IPrincipalProxy]
infrastructure/users_stack.py:32: note:     Got:
infrastructure/users_stack.py:32: note:         def __jsii_proxy_class__() -> Type[_PrincipalBaseProxy]

```

**Review by l0b0 — APPROVED** [review]

**Review by billgeo — APPROVED** [review]
